### PR TITLE
poc

### PR DESCRIPTION
**Pull request by jrcamp**

### CODE REVIEW

File: etl-sdk/src/main/java/io/ndustrial/etl/config/Builder.kt [hunk 1]
```diff
@@ -83,20 +84,27 @@ abstract class ConfigBuilder<T>(val clazz: Class<T>,
 /**
  * Source for retrieving a resource.
  */
-typealias Source = () -> String
+abstract class Source(val path: String) {
```
Comment (borregoho) [inline etl-sdk/src/main/java/io/ndustrial/etl/config/Builder.kt]: I think other than the package name qualification a more descriptive/nemonic name would be beneficial, like something that suggests this is about ETL and Config/Yaml fetching

Comment (borregoho) [reply etl-sdk/src/main/java/io/ndustrial/etl/config/Builder.kt]: I think a simpler interface would be preferred, one that makes no assumption about the underlying implementation of the concrete resources fetcher, like the one I had before:

```
interface EtlResourcesProvider {
    /**
     * Fetches an etl resource content given its path
     * (currently only etl plans are supported)
     */
    fun getResourceContent(resourcePath: String): String
}
```

or something similar, single method, takes resource id/path and returns the resource content as a string, plain and simple.

this Source interface is implying a File System like hierarchy, forcing a sibling relationship between the plan resources, and a parent/child relationship on orchestration resource and its etl plans. That sounds like abstraction leaking, those specifics should live in the concrete implementations side. Just think about the management API, which would be another implementation, I don't see how it fits in the protocol that this interface forces.

File: etl-sdk/src/main/java/io/ndustrial/etl/config/EtlOrchestrationConfig.kt [hunk 2]
```diff
@@ -36,13 +35,14 @@ class EtlOrchestrationConfig {
...
 
-    fun init(provider: EtlResourcesProvider): EtlOrchestrationConfig {
+    fun init(provider: Source): EtlOrchestrationConfig {
         LOG.info("Initializing $this")
 
         LOG.info("Reading EtlPlans: $planResources")
 
         plans = planResources.map {
-            EtlPlanBuilder().fromSource { provider.getResourceContent(it) }
+            require(!Paths.get(it).isAbsolute, { "$it must be a relative path" })
```
Comment (borregoho) [inline etl-sdk/src/main/java/io/ndustrial/etl/config/EtlOrchestrationConfig.kt]: abstraction leaking

File: etl-sdk/src/main/java/io/ndustrial/etl/config/EtlOrchestrationConfig.kt [hunk 3]
```diff
@@ -36,13 +35,14 @@ class EtlOrchestrationConfig {
...
+            EtlPlanBuilder().fromSource(provider.resolveSibling(it))
```
Comment (borregoho) [inline etl-sdk/src/main/java/io/ndustrial/etl/config/EtlOrchestrationConfig.kt]: the concept of sibling is very restrictive, and IMHO unnecessary

File: etl-sdk/src/main/java/io/ndustrial/etl/config/EtlOrchestrationConfig.kt [hunk 4]
```diff
@@ -84,27 +66,14 @@ class EtlOrchestrationConfig {
     }
 }
 
-interface EtlResourcesProvider {
```
Comment (borregoho) [inline etl-sdk/src/main/java/io/ndustrial/etl/config/EtlOrchestrationConfig.kt]: still would like to be able to use this interface

File: etl-sdk/src/main/java/io/ndustrial/etl/config/Builder.kt (hunk 1 above)
Comment (borregoho) [reply etl-sdk/src/main/java/io/ndustrial/etl/config/Builder.kt]: furthermore, I really like the fact that both the class path source and the GitHub source can parse the same resource structure (which is not relative path based). Wonder how will that work with this design

**Comment by jrcamp** [comment]
I don't necessarily disagree but I wasn't intending this change to be used as-is but for you to incorporate the idea into your changes. Basically resource loading should be decoupled from ConfigBuilder. This was just showing one way to do it.

**Pull request closed without merging**